src/scenic/core/visibility.py

```diff
@@ -0,0 +1,503 @@
+                    vertices = [position.coordinates] + vertices
+                    lines = [trimesh.path.entities.Line([0,v]) for v in range(1,len(vertices))]
+                    occluded_colors = [(255,0,0,255) for line in occluded_vertices]
+                    clear_colors = [(0,255,0,255) for line in clear_vertices]
+                    colors = occluded_colors + clear_colors
+                    render_scene = trimesh.scene.Scene()
+                    render_scene.add_geometry(trimesh.path.Path3D(entities=lines, vertices=vertices, process=False, colors=colors))
+                    render_scene.add_geometry(target.occupiedSpace.mesh)
+                    render_scene.add_geometry(occ_obj.occupiedSpace.mesh)
+                    render_scene.show()
+
+            if len(candidate_rays) > 0:
+                return True
+
+        # No rays hit the object and are not occluded, so the object is not visible
+        return False
+
+    elif isinstance(target, (Point, OrientedPoint, Vector)):
+        if isinstance(target, (Point, OrientedPoint)):
+            target_loc = target.position
+        else:
+            target_loc = target
+
+        # First check if the distance to the point is less than or equal to the visible distance. If not, the object cannot
+        # be visible.
+        target_distance = position.distanceTo(target_loc)
+        if target_distance > visibleDistance:
+            return False
+
+        # Create the single candidate ray and check that it's within viewAngles.
+        target_vertex = target_loc - position
+        candidate_ray = target_vertex/np.linalg.norm(target_vertex)
+
+        azimuth = np.mod(np.arctan2(candidate_ray[1], candidate_ray[0]) - math.pi/2 + np.pi, 2*np.pi) - np.pi
+        altitude = np.arcsin(candidate_ray[2]/(np.linalg.norm(candidate_ray)))
+
+        if orientation is not None:
+            relative_azimuth = azimuth - orientation.yaw
+            relative_altitude = altitude - orientation.pitch
+        else:
+            relative_azimuth = azimuth
+            relative_altitude = altitude
+
+        candidate_ray_list = np.array([candidate_ray])
+
+        ## DEBUG ##
+        #Show all original candidate rays
+        if debug:
+            vertices = [visibleDistance*vec + position.coordinates for vec in candidate_ray_list]
+            vertices = [position.coordinates] + vertices
+            lines = [trimesh.path.entities.Line([0,v]) for v in range(1,len(vertices))]
+            colors =[(0,0,255,255) for line in lines]
+
+            render_scene = trimesh.scene.Scene()
+            render_scene.add_geometry(trimesh.path.Path3D(entities=lines, vertices=vertices, process=False, colors=colors))
+            render_scene.add_geometry(SpheroidRegion(position=target_loc, dimensions=(0.1,0.1,0.1)))
+            for i in list(occludingObjects):
+                render_scene.add_geometry(i.occupiedSpace.mesh)
+            render_scene.show()
```

Comment:
This visualization code keeps getting repeated and should be factored into a separate function.

Reply:
Agreed, I put it on the todo list as part of the refactor.